src/sentry/scripts/quotas/is_rate_limited.lua

```diff
@@ -1,39 +1,46 @@
 -- Check a collection of quota counters to identify if an item should be rate
 -- limited. Values provided as ``KEYS`` specify the keys of the counters to
--- check, and values provided as ``ARGV`` specify the maximum value (quota
--- limit), expiration time for each key, and a key for the corresponding
--- counter to subtract.
+-- check and the keys of counters to subtract, and values provided as ``ARGV``
+-- specify the maximum value (quota limit) and expiration time for each key.
 --
--- For example, to check a quota ``foo`` that has a limit of 10 items and
--- expires at the Unix timestamp ``100``, as well as a quota ``bar`` that has a
--- limit of 20 items and should expire at the Unix timestamp ``100``, the
--- ``KEYS`` and ``ARGV`` values would be as follows:
+-- For example, to check a quota ``foo`` that has a corresponding refund/negative
+-- counter "subtract_from_foo", a limit of 10 items and expires at the Unix timestamp
+-- ``100``, as well as a quota ``bar`` that has a corresponding refund/negative
+-- counter "subtract_from_bar" limit of 20 items and should expire at the Unix
+-- timestamp ``100``, the ``KEYS`` and ``ARGV`` values would be as follows:
 --
---   KEYS = {"foo", "bar"}
---   ARGV = {10, 100, "baz", 20, 100, "foo"}
+--   KEYS = {"foo", "subtract_from_foo", "bar", "subtract_from_bar"}
+--   ARGV = {10, 100, 20, 100}
 --
 -- If all checks pass (the item is accepted), the counters for all quotas are
 -- incremented. If any checks fail (the item is rejected), the counters for all
 -- quotas are unaffected. The result is a Lua table/array (Redis multi bulk
 -- reply) that specifies whether or not the item was *rejected* based on the
 -- provided limit.
-assert(#KEYS * 3 == #ARGV, "incorrect number of keys and arguments provided")
+assert(#KEYS == #ARGV, "incorrect number of keys and arguments provided")
+assert(#KEYS % 2 == 0, "there must be an even number of keys")
 
 local results = {}
 local failed = false
+local results_i = 1
 for i=1,#KEYS do
-    local limit = tonumber(ARGV[(i * 3) - 2])
-    local rejected = (redis.call('GET', KEYS[i]) or 0) - (redis.call('GET', ARGV[(i * 3)]) or 0) + 1 > limit
-    if rejected then
-        failed = true
+    if i % 2 ~= 0 then
```

Comment:
In my opinion, this is simpler if you use a step value of 2 in the for loop statement, e.g.

```lua
> for i = 1, 8, 2 do
>> print(i)
>> end
1
3
5
7
```

You can then use:

- `i` as the quota key index (`KEYS[i]`),
- `i + 1` as the refund index (`KEYS[i+1]`),
- `i` as the limit value (`ARGV[i]`), and
- `i + 1` as the expiration value (`ARGV[i + 1]`)
- `i + 1 / 2` as the result iteration counter rather than introducing a new local variable.